src/data_hub/dataframe.py

```diff
@@ -0,0 +1,159 @@
+
+    @property
+    def at(self):
+        """Override the `at` property to return an AtIndexer"""
+        return self.AtIndexer(self)
+
+    def __setitem__(self, key, value):
+        """Override the __setitem__ method to update the Deep Origin database when changes are made to the local
+        dataframe"""
+
+        # first, call the pandas method
+        super().__setitem__(key, value)
+
+        # now, update the Deep Origin database with the changes
+        # we just made
+        if self.auto_sync:
+            self.sync(columns=[key])
+
+    def _repr_html_(self):
+        """method override to customize printing in a Jupyter notebook"""
+
+        name = self.attrs["metadata"]["name"]
+        url = construct_resource_url(
+            name=name,
+            row_type="database",
+        )
+
+        # Convert the string to a datetime object
+        date_str = self.attrs["metadata"]["dateCreated"]
+        date_obj = datetime.strptime(date_str, "%Y-%m-%d %H:%M:%S.%f")
+
+        now = datetime.now()
+
+        # Calculate the difference
+        time_diff = now - date_obj
+
+        # Convert the time difference into "x time ago" format
+        time_ago = humanize.naturaltime(time_diff)
+
+        header = f'<h4>{name} <a href = "{url}">🔗</a></h4>'
+        txt = f'<p style="font-size: 12px; color: #808080;">Created {time_ago}.</p>'
+        df_html = super()._repr_html_()
+        return header + txt + df_html
+
+    def __repr__(self):
+        """method override to customize printing in an interactive session"""
+
+        header = f'{self.attrs["metadata"]["hid"]}\n'
+        df_representation = super().__repr__()
+        return header + df_representation
+
+    def sync(
+        self,
+        *,
+        columns: Optional[list] = None,
+        rows: Optional[list] = None,
+    ):
+        """Manually synchronize data in the dataframe to the underlying Deep Origin database.
+
+        !!! tip "Deep Origin DataFrames automatically synchronize"
+            Typically, you do not need to manually synchronize. If the `auto_sync` attribute of the dataframe is set to `True`, the dataframe will automatically synchronize when changes are made to the dataframe.
+
+
+        Args:
+            columns (list, optional): The columns of the dataframe to update. Defaults to None.
+            rows (list, optional): The rows to update. Defaults to None. When None, all rows in the relevant columns are updated.
+
+        """
+
+        if columns is None:
+            columns = self.columns
+
+        for column in columns:
+            if column in ["Validation Status", "ID"]:
+                continue
+
+            column_metadata = [
+                col for col in self.attrs["metadata"]["cols"] if col["name"] == column
+            ]
+
+            if len(column_metadata) == 0:
+                raise NotImplementedError(
+                    "Column metadata not found. This is likely because it's a new column"
+                )
```

Comment:
In _theory_ could we implement automatic column creation?

Reply:
yeah. a little tricky because we need to guess types. but possible. i'll make a ticket for this

Reply:
https://deeporigin.atlassian.net/browse/DOS-3660